test/test_heat.py

```diff
@@ -0,0 +1,241 @@
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import pyopencl.array as cla  # noqa
+import pyopencl.clmath as clmath # noqa
+from pytools.obj_array import make_obj_array
+import pymbolic as pmbl
+import pymbolic.primitives as prim
+import mirgecom.symbolic as sym
+from mirgecom.heat import heat_operator
+from meshmode.dof_array import thaw, flat_norm
+
+from meshmode.array_context import (  # noqa
+    pytest_generate_tests_for_pyopencl_array_context
+    as pytest_generate_tests)
+
+import pytest
+import os
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+# Tests below take a problem description as input, which is a tuple
+#   (dim, alpha, mesh_factory, sym_u)
+# where:
+#   dim is the problem dimension
+#   alpha is the diffusivity
+#   mesh_factory is a factory that creates a mesh given a characteristic size
+#   sym_u is a symbolic expression for the solution
+
+
+def get_decaying_cosine(dim, alpha):
+    # 1D: u(x,t) = exp(-alpha*t)*cos(x)
+    # 2D: u(x,y,t) = exp(-2*alpha*t)*cos(x)*cos(y)
+    # 3D: u(x,y,z,t) = exp(-3*alpha*t)*cos(x)*cos(y)*cos(z)
+    # on [-pi/2, pi/2]^{#dims}
+    def mesh_factory(n):
+        from meshmode.mesh.generation import generate_regular_rect_mesh
+        return generate_regular_rect_mesh(
+            a=(-0.5*np.pi,)*dim,
+            b=(0.5*np.pi,)*dim,
+            n=(n,)*dim)
+    sym_coords = prim.make_sym_vector("x", dim)
+    sym_t = pmbl.var("t")
+    sym_cos = pmbl.var("cos")
+    sym_exp = pmbl.var("exp")
+    sym_u = sym_exp(-dim*alpha*sym_t)
+    for i in range(dim):
+        sym_u *= sym_cos(sym_coords[i])
+    return (dim, alpha, mesh_factory, sym_u)
+
+
+def sym_heat(dim, sym_u):
+    """Return symbolic expressions for the heat equation system given a desired
+    solution. (Note: In order to support manufactured solutions, we modify the heat
+    equation to add a source term (f). If the solution is exact, this term should
+    be 0.)
+    """
+
+    sym_alpha = pmbl.var("alpha")
+    sym_t = pmbl.var("t")
+
+    # rhs = alpha * div(grad(u))
+    sym_rhs = sym_alpha * sym.div(sym.grad(dim, sym_u))
+
+    # f = u_t - rhs
+    sym_f = sym.diff(sym_t)(sym_u) - sym_rhs
+
+    return sym_f, sym_rhs
+
+
+# Note: Must integrate in time for a while in order to acheive expected spatial
+# accuracy. Checking the RHS alone will give lower numbers.
```

Comment:
```suggestion
# Note: Must integrate in time for a while in order to acheive expected spatial
# accuracy. Checking the RHS alone will give lower numbers.
#
# Working hypothesis: RHS lives in lower order polynomial space and thus doesn't
# attain full-order convergence.
```
(My main point is that we preserve our thought process in the code, to justify why it's there. It may well turn out to be wrong, that's OK.)